hw/ssi/hercules_spi.c

```diff
@@ -0,0 +1,723 @@
+    SPIPC0          = 0x014,
+    ENAFUN          = BIT(8),
+    SPIPC1          = 0x018,
+    SPIPC8          = 0x034,
+    SPIDAT0         = 0x038,
+    SPIDAT1         = 0x03C,
+    SPIBUF          = 0x040,
+    SPIFMT0         = 0x050,
+    SPIFMT1         = 0x054,
+    SPIFMT2         = 0x058,
+    SPIFMT3         = 0x05C,
+    MIBSPIE         = 0x070,
+    RXRAM_ACCESS    = BIT(16),
+    TGINTFLAG       = 0x084,
+    LTGPEND         = 0x094,
+    TG0CTRL         = 0x098,
+    TG14CTRL        = 0x0D0,
+    TG15CTRL        = 0x0D4,
+    DMA0CTRL        = 0x0DF,
+    RXDMAENA        = BIT(15),
+    TXDMAENA        = BIT(14),
+    DMA8CTRL        = 0x0F4,
+    DMA0COUNT       = 0x0F8,
+    DMA8COUNT       = 0x112,
+    DMACNTLEN       = 0x118,
+    TGENA           = BIT(31),
+    ONESHOTx        = BIT(30),
+    PAR_ECC_CTRL    = 0x120,
+    PAR_ECC_STAT    = 0x124,
+    UERR_FLG0       = BIT(0),
+    UERR_FLG1       = BIT(1),
+    SBE_FLG0        = BIT(8),
+    SBE_FLG1        = BIT(9),
+    UERRADDR1       = 0x128,
+    UERRADDR0       = 0x12C,
+    IOLPBKTSTCR     = 0x134,
+    IOLPBKSTENA     = 0x0A,
+    ECCDIAG_CTRL    = 0x140,
+    ECCDIAG_STAT    = 0x144,
+    SEFLG0          = BIT(0),
+    SEFLG1          = BIT(1),
+    DEFLG0          = BIT(16),
+    DEFLG1          = BIT(17),
+    SBERRADDR1      = 0x148,
+    SBERRADDR0      = 0x14C,
+};
+
+#define INTFLGRDY(n)               BIT((n) + 16)
+
+#define TXRAM_CSNR(w)              extract32(w, 16, 8)
+#define TXRAM_DFSEL(w)             extract32(w, 24, 2)
+#define TXRAM_TXDATA(w, l)         extract32(w, 0, l)
+#define TXRAM_CSHOLD               BIT(28)
+
+#define SPIFMT_CHARLEN(w)          extract32(w, 0, 5)
+
+#define TGxCTRL_PSTART(w)          extract32(w, 8, 8)
+
+#define IOLPBKTSTCR_IOLPBKSTENA(w) extract32(w, 8, 4)
+
+#define TXDMA_MAPx(w)              extract32(w, 16, 4)
+#define RXDMA_MAPx(w)              extract32(w, 20, 4)
+
+#define SBE_EVT_EN(w)              extract32(w, 24, 4)
+
+static uint16_t hercules_spi_tx_single(HerculesMibSpiState *s, uint32_t txword)
+{
+    const uint8_t  dfsel   = TXRAM_DFSEL(txword);
+    const uint32_t spifmt  = s->spifmt[dfsel];
+    const uint8_t  charlen = SPIFMT_CHARLEN(spifmt);
+    const uint16_t txdata  = TXRAM_TXDATA(txword, charlen);
+
+    /*
+     * FIXME: Should probably cache this as a pivate boolean flag
+     */
+    if (unlikely(IOLPBKTSTCR_IOLPBKSTENA(s->iolpbktstcr) == IOLPBKSTENA)) {
+        return txdata;
+    }
+
+    return ssi_transfer(s->ssi, txdata);
+}
+
+static bool hercules_spi_ram_big_endian(HerculesMibSpiState *s)
+{
+    return s->io.regs.ops->endianness == DEVICE_BIG_ENDIAN;
+}
+
+static uint32_t hercules_spi_ram_read(HerculesMibSpiState *s, uint32_t *word)
+{
+    return hercules_spi_ram_big_endian(s) ? ldl_be_p(word) : ldl_le_p(word);
+}
+
+static void hercules_spi_ram_write(HerculesMibSpiState *s, uint32_t *word,
+                                   uint32_t value)
+{
+    if (hercules_spi_ram_big_endian(s)) {
```

Comment:
Another spot that we need to check on RM57

Reply:
yea... I dug into this a big more and will comment separately

Reply:
I at least checked that spi transfers complete in an expected way without the external chardev doing much (you can setup/start transfers)